application/common/models/User.php

```diff
@@ -0,0 +1,190 @@
+
+        $scenarios[self::SCENARIO_AUTHENTICATE] = ['username', 'password'];
+
+        return $scenarios;
+    }
+
+    public function rules(): array
+    {
+        return ArrayHelper::merge([
+            [
+                'active', 'default', 'value' => 'yes',
+            ],
+            [
+                'locked', 'default', 'value' => 'no',
+            ],
+            [
+                ['active', 'locked'], 'in', 'range' => ['yes', 'no'],
+            ],
+            [
+                'email', 'email',
+            ],
+            [
+                'password', 'required',
+                'on' => [self::SCENARIO_UPDATE_PASSWORD, self::SCENARIO_AUTHENTICATE],
+            ],
+            [
+                'password', 'string',
+            ],
+            [
+                'password',
+                $this->validatePassword(),
+                'on' => self::SCENARIO_AUTHENTICATE,
+            ],
+            [
+                ['last_synced_utc', 'last_changed_utc'],
+                'default', 'value' => Utils::now(),
+            ],
+            [
+                ['employee_id', 'first_name', 'last_name', 'display_name', 'username', 'email'],
+                'string',
+            ],
+        ], parent::rules());
+    }
+
+    private function validatePassword(): \Closure
+    {
+        return function ($attributeName) {
+            if (! password_verify($this->$attributeName, $this->password_hash)) {
+                $this->addError($attributeName, 'Incorrect password.');
+            }
+        };
+    }
+
+    public function behaviors(): array
+    {
+        return [
+            'changeTracker' => [
+                'class' => AttributeBehavior::className(),
+                'attributes' => [
+                    ActiveRecord::EVENT_BEFORE_INSERT => 'last_changed_utc',
+                    ActiveRecord::EVENT_BEFORE_UPDATE => 'last_changed_utc',
+                ],
+                'value' => Utils::now(),
+                'skipUpdateOnClean' => true, // only update the column if the model is dirty
+            ],
+            'updateTracker' => [
+                'class' => AttributeBehavior::className(),
+                'attributes' => [
+                    ActiveRecord::EVENT_BEFORE_INSERT => 'last_synced_utc',
+                    ActiveRecord::EVENT_BEFORE_UPDATE => 'last_synced_utc',
+                ],
+                'value' => Utils::now(),
+                'skipUpdateOnClean' => false, // always update the column regardless of dirtiness
+            ],
+        ];
+    }
+
+    /**
+     * @return array of fields that should be included in responses.
+     */
+    public function fields(): array
+    {
+        return [
+            'employee_id',
+            'first_name',
+            'last_name',
+            'display_name',
+            'username',
+            'email',
+            'active',
+            'locked',
+            'last_changed_utc',
+            'last_synced_utc',
```

Comment:
We probably need to check with @fillup about whether to include the changed/synced fields in responses. I left them out of the RAML spec for the API because I still don't know of any use case for sending them, so we'll just need to figure out which way we want to go.